KeyPolicy: add custom constructor and make all fields private

aarongable opened the pull request:
Change how goodkey.KeyPolicy keeps track of allowed RSA and ECDSA key sizes, to make it slightly more flexible while still retaining the very locked-down allowlist of only 6 acceptable key sizes (RSA 2048, 3076, and 4092, and ECDSA P256, P384, and P521). Add a new constructor which takes in a collection of allowed key sizes, so that users of the goodkey package can customize which keys they accept. Rename the existing constructor to make it clear that it uses hardcoded default values.

With these new constructors available, make all of the goodkey.KeyPolicy member fields private, so that a KeyPolicy can only be built via these constructors.

~~NOTE: This PR is based on top of https://github.com/letsencrypt/boulder/pull/7541. DO NOT MERGE until that PR has landed.~~

pgporada commented on goodkey/good_key_test.go:
```diff
@@ -289,6 +290,17 @@ func TestDBBlocklistReject(t *testing.T) {
 
+func TestDefaultECDSAKeys(t *testing.T) {
+	policy, err := NewDefaultKeyPolicy(&Config{}, nil)
+	test.AssertNotError(t, err, "NewKeyPolicy failed")
```
```suggestion
	test.AssertNotError(t, err, "NewDefaultKeyPolicy failed")
```

pgporada commented on goodkey/good_key_test.go:
```diff
@@ -289,6 +290,17 @@ func TestDBBlocklistReject(t *testing.T) {
 	test.AssertEquals(t, err.Error(), "public key is forbidden")
 }
 
+func TestDefaultECDSAKeys(t *testing.T) {
```
This should be renamed to something like, TestDefaultAllowedKeys.

aarongable commented:
Apologies for the force-push; the review diff is still acting as though this PR _adds_ support for P521, despite the merge of the upstream branch. I've rebased and cleaned it up and the review diff is now much more accurate.

jsha reviewed (changes requested):
We have a `goodkey.Config` struct. Why not make `AllowedKeys` part of that struct?

If we want to get the defaulting behavior, we should have a `*AllowedKeys` field of the `Config` struct, and have `nil` mean "Let's Encrypt defaults". Alternately: we could have a `const LetsEncryptCPS AllowedKeys`. That would make every call site that references the constant really clear that this is a CPS-related value.

Two more things that are really pre-existing issues:

 - Since `config *Config` is a pointer, the zero value (`nil`) should be useful, and it should mean "supply default values."
 - `NewCustomKeyPolicy` and `NewDefaultKeyPolicy` both repeat "key" from the package name. We could simplify to just `goodkey.New` and `goodkey.NewCustom` (or `goodkey.NewPolicy` and `goodkey.NewCustomPolicy` if we want to be more verbose).

aarongable commented:
> We have a `goodkey.Config` struct. Why not make `AllowedKeys` part of that struct?

My thinking was that I wanted to be really sure that we couldn't accidentally configure Boulder to accept key types not allowed by our CPS. To that end, I kept the AllowedKeys out of the JSON config, and required a different method with a different method signature to get access to non-default sets of keys.

I can definitely simplify it though, if that still feels safe enough to you. I'll upload a commit showing that simplification, and then we can back it out if it feels like riding too close to the edge.

jsha commented:
Ah, I missed the context that `Config` was for reading from JSON and thought it was only for in-memory configuration.

I think it's probably okay to add to the JSON config. Maybe we should have guardrails in e.g. the CA that checks whether its provided `goodkey.KeyPolicy` object has no more than the CPS-allowed types enabled? But that would require making those fields public again... hm.

aarongable commented:
I've added a commit which moves AllowedKeys inside the Config struct, creates a LetsEncryptCPS AllowedKeys const (well, really a func, because we can't do const structs in go), and simplifies and renames the constructors. Let me know what you think, I'm happy to move forward this way or roll back to the previous version.

jsha reviewed (approved):
Looks good!
